modules/events.js

```diff
@@ -32,286 +32,403 @@
+            delete this._events[type];
+        }
+    } else if (list === listener ||
+             list.listener && list.listener === listener) {
+        delete this._events[type];
     }
 
-    if (position < 0) return this;
-    list.splice(position, 1);
-    if (list.length === 0)
-      delete this._events[type];
-  } else if (list === listener ||
-             (list.listener && list.listener === listener))
-  {
-    delete this._events[type];
-  }
-
-  return this;
+    return this;
 };
 
+/**
+ * Removes all listeners, or those of the specified event.
+ *
+ * @param  String    type      An event name
+ * @return EventEmitter        A reference to the EventEmitter so calls can be chained
+ */
 EventEmitter.prototype.removeAllListeners = function removeAllListeners(type) {
-  if (arguments.length === 0) {
-    this._events = {};
-    return this;
-  }
+    "use strict";
+    if (arguments.length === 0) {
+        this._events = {};
+        return this;
+    }
 
-  // does not use listeners(), so no side effect of creating _events[type]
-  if (type && this._events && this._events[type]) this._events[type] = null;
-  return this;
+    // does not use listeners(), so no side effect of creating _events[type]
+    if (type && this._events && this._events[type]) {
+        this._events[type] = null;
+    }
+    return this;
 };
 
+/**
+ * get callbacks listener.
+ *
+ * @param  String    type      An event name
+ * @return mixed
+ */
 EventEmitter.prototype.listeners = function listeners(type) {
-  if (!this._events) this._events = {};
-  if (!this._events[type]) this._events[type] = [];
-  if (!isArray(this._events[type])) {
-    this._events[type] = [this._events[type]];
-  }
-  return this._events[type];
+    "use strict";
+    if (!this._events) {
+        this._events = {};
+    }
+    if (!this._events[type]) {
+        this._events[type] = [];
+    }
+    if (!isArray(this._events[type])) {
+        this._events[type] = [this._events[type]];
+    }
+    return this._events[type];
 };
 
-// Added for CasperJS: filters a value attached to an event
+/**
+ * Trigger filter attached to an event and return value(s).
+ *
+ * @param  String    type      An event name
+ * @param  Array     args...  The rest of arguments passed to fn
+ * @return mixed
+ */
 EventEmitter.prototype.filter = function filter() {
-  var type = arguments[0];
-  if (!this._filters) {
-    this._filters = {};
-    return;
-  }
-
-  var _filter = this._filters[type];
-  if (typeof _filter !== 'function') {
-    return;
-  }
-  return _filter.apply(this, Array.prototype.splice.call(arguments, 1));
+    "use strict";
+    var type = arguments[0], res;
+    if (!this._filters) {
+        this._filters = {};
+    }
+    var _filter = this._filters[type];
+    if (typeof _filter === 'function') {
+        res = _filter.apply(this, Array.prototype.slice.call(arguments, 1));
+    }
+    this.emit(Array.prototype.slice.call(arguments, 0).concat([res]));
```

Comment:
this only line has been added :  
   this.emit(Array.prototype.slice.call(arguments, 0).concat([res]));